charmrepo/charmstore.go

```diff
@@ -63,11 +69,65 @@ func NewCharmStore(p NewCharmStoreParams) (Interface, error) {
 	}, nil
 }
 
-var notImplemented = errgo.New("not implemented")
-
 // Get implements Interface.Get.
 func (s *CharmStore) Get(curl *charm.URL) (charm.Charm, error) {
-	return nil, notImplemented
+	// Disable bundle deployment.
+	if curl.Series == "bundle" {
+		return nil, errgo.Newf("expected a charm URL, got bundle URL %q", curl)
+	}
+
+	// Prepare the cache directory and retrieve the charm.
+	if err := os.MkdirAll(s.cacheDir, os.FileMode(0755)); err != nil {
+		return nil, errgo.Notef(err, "cannot create the cache directory")
+	}
+	r, id, expectHash, _, err := s.client.GetArchive(curl.Reference())
```

Comment:
I think you should use the size here and check that the number of bytes returned by io.Copy is as expected. An error message that says there's a size mismatch, giving expected and actual sizes, is going to be much more accessible than a hash mismatch message (which of course needs to be done too).